zookeeper-server/src/test/java/org/apache/zookeeper/server/backup/storage/impl/FileSystemBackupStorageTest.java

```diff
@@ -0,0 +1,144 @@
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.FixMethodOrder;
+import org.junit.Test;
+import org.junit.runners.MethodSorters;
+
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class FileSystemBackupStorageTest {
+  private static String testFolderPath;
+  private static String testFileDirPath;
+  private static String backupFileDirPath;
+  private static List<String> testFileNames =
+      new ArrayList<>(Arrays.asList("log.testFile_0", "snapshot.testFile_1", "log.testFile_2"));
+  private static BackupStorageProvider backupStorage;
+  private static List<File> testFiles;
+  private static List<File> backupFiles;
+  private static final String NAMESPACE = "test";
+
+  @BeforeClass
+  public static void beforeClass() throws ConfigException, IOException {
+    // Build backup config that point the backup storage to a local directory
+    BackupConfig backupConfig = new BackupConfig.Builder().setEnabled(true)
+        .setStatusDir(new File("./localBackupStorageTest/backup/status")).
+            setTmpDir(new File("./localBackupStorageTest/tmp/backup"))
+        .setStorageProviderClassName(FileSystemBackupStorage.class.getName())
+        .setBackupStoragePath("./localBackupStorageTest").setNamespace(NAMESPACE).build().get();
+
+    backupStorage = new FileSystemBackupStorage(backupConfig);
+    testFolderPath =
+        Paths.get(Paths.get("").toAbsolutePath().toString(), backupConfig.getBackupStoragePath())
+            .toString();
+    backupFileDirPath = String.join(File.separator, testFolderPath, NAMESPACE);
+    testFileDirPath = String.join(File.separator, testFolderPath, "testFiles");
+    testFiles = new ArrayList<>();
+    backupFiles = new ArrayList<>();
+
+    for (String testFileName : testFileNames) {
+      String testFilePath = String.join(File.separator, testFileDirPath, testFileName);
+      File testFile = new File(testFilePath);
+      testFiles.add(testFile);
+      BackupStorageUtil.createFile(testFile, false);
+      FileWriter fileWriter = new FileWriter(testFile);
+      for (int i = 0; i < 1000; i++) {
+        fileWriter.write(testFileName);
+        fileWriter.write('\n');
+      }
+      fileWriter.close();
+      Assert.assertTrue(testFile.exists());
+      Assert.assertTrue(sizeOf(testFile) > 0);
+    }
+  }
+
+  @AfterClass
+  public static void afterClass() throws IOException {
+    FileUtils.deleteDirectory(new File(testFolderPath));
+  }
+
+  @Test
+  public void test0_CopyToBackupStorage() throws IOException {
+    for (File testFile : testFiles) {
+      backupStorage.copyToBackupStorage(testFile, testFile);
+      String backupFilePath =
+          BackupStorageUtil.constructBackupFilePath(testFile.getName(), backupFileDirPath);
+      File backupFile = new File(backupFilePath);
+      backupFiles.add(backupFile);
+      Assert.assertEquals(sizeOf(testFile), sizeOf(backupFile));
```

Comment:
Curious, is there a reason we're doing a size comparison instead of the content comparison?

Reply:
This comparison is meant to make sure the data is streamed correctly that there is no data loss during the process, therefore I used the size comparison. Regarding to the content of the file, the file content is arbitrarily created with repeating the file name, so comparing the content doesn't provide much more value than the size comparison in my opinion. If we want to do content comparison, we need to read the content into strings and compare two long strings, which is a heavy operation.

Reply:
Sounds fair.